CONTRIBUTING.md

```diff
@@ -0,0 +1,77 @@
+* If you are adding new code, please include proper documentation outside of code comments where it makes sense. Ideally, for adding new features, it helps to at the very minimum add a `README.md` which explains purpose, inner workings, etc. In some cases, it may also be helpful to include UML diagrams which depict the new feature/component (sequence diagrams, component diagrams and workflows are often good choices).
+* Use the `context` package to manage goroutine lifetimes. For good examples, look at how the `Chain` and `Mempool` are managed using contexts.
+
+### Performance
+
+* Wherever possible, when allocating slices, try to allocate the entire slice in one go, and then assigning to the indices (or using `copy` where applicable), instead of creating an empty slice and then calling `append`. This prevents an unnecessary amount of dynamic memory allocations.
+* When creating a channel which sole purpose is to send signals, and not data, use a `chan struct{}`. The empty struct in Go takes up 0 bytes of space, and is thus the most efficient way of sending signals.
+* When clearing a map, use a `range` loop, and delete every key. This is more efficient than replacing the map with a new instance, as the GC does not have to clean this up later.
+
+### Committing
+
+* Rebase your commits generously. Avoid having to open pull requests with a massive amount of commits, which all essentially work towards the same issue.
+* If possible, try to include all of the code, tests and documentation for one issue, into one commit.
+* The commit title should be in imperative form. Examples:
+  * Implement ...
+  * Add ...
+  * Change ...
+* The commit title should only state what you did. This helps keep commit titles short and easy to find.
+* The commit description should explain why you did it. Most times, referring to the issue itself is enough. Please do so with keywords (`fixes`, `resolves` etc.) in order to automatically link issues to commits. This helps to retrace commits and it simplifies opening pull requests.
+* Do not elaborate on how you did what you did - the code should be self-explanatory in this case. If the code is not self-explanatory enough (due to performance shortcuts or the sort), please explain this in comments, rather than in the commit.
+
+### Testing
+
+* One test should test one thing. Test code is still code, and so we should try to separate concerns and make it as readable as possible. This rule may be broken in cases where it does not make sense to separate concerns.
+* If there are internal and external tests related to the package, then mark the files with name\_in\_ test.go and name\_ex\_test.go
+* Make use of the `testify` package where possible. This clearly separates test code from production code, since the error handling is noticably different. It also provides useful functionality for checking equality, making sure certain fields are empty/not empty etc.
+
```

Comment:
I would also add a couple of bullet points about table testing, the need to instantiate `require`/`assert` when multiple statements get tested, and recommendation on when to use a `TestMain`